From c038ca07ccd599a568bb1a9901742d2197184a03 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Sat, 25 Jul 2015 00:38:11 +0200
Subject: [PATCH 1/2] Workaround for invisible icons on GTK+ 3

- Fix background drawing.
- Button label size hack is not needed on GTK+ 3.
- Disable monitoring changes in menu on GTK+ 3, as it currently causes invisible icons.
---
 src/lxlauncher.c | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/src/lxlauncher.c b/src/lxlauncher.c
index 17f91f9..b6f566e 100644
--- a/src/lxlauncher.c
+++ b/src/lxlauncher.c
@@ -197,6 +197,7 @@ static GtkWidget* add_btn( GtkWidget* table, const char* text, GdkPixbuf* icon,
 
     gtk_widget_set_tooltip_text(btn, tip);
 
+#if !GTK_CHECK_VERSION(3, 0, 0)
     /* Adjust the size of label and set line wrapping is needed.
      * FIXME: this is too dirty, and the effect is quite limited.
      * However, due to the unfortunate design flaws of gtk+, the
@@ -223,11 +224,6 @@ static GtkWidget* add_btn( GtkWidget* table, const char* text, GdkPixbuf* icon,
     }
 
     gtk_widget_set_app_paintable( btn, TRUE );
-#if GTK_CHECK_VERSION(3, 0, 0)
-    GdkRGBA color;
-    gdk_rgba_parse(&color, "none");
-    gtk_widget_override_background_color ((GtkWidget*)btn, GTK_STATE_FLAG_NORMAL, &color);
-#else
     gdk_window_set_back_pixmap( ((GtkWidget*)btn)->window, NULL, TRUE );
 #endif
     return btn;
@@ -291,6 +287,9 @@ static gboolean on_viewport_draw( GtkWidget* w, cairo_t *cr1, gpointer data )
         {
             pixbuf = gdk_pixbuf_scale_simple(pixbuf, alloc.width, alloc.height,
                                              GDK_INTERP_BILINEAR);
+            GdkRGBA color;
+            gdk_rgba_parse(&color, "none");
+            gtk_widget_override_background_color (w, GTK_STATE_FLAG_NORMAL, &color);
             g_object_set_data_full(G_OBJECT(w), "LXLauncher:background",
                                    pixbuf, g_object_unref);
         }
@@ -989,7 +988,10 @@ int main(int argc, char** argv)
 #else
     root_dir = MENU_CACHE_DIR(menu_cache_item_ref(MENU_CACHE_ITEM(menu_cache_get_root_dir( menu_tree ))));
 #endif
+#if !GTK_CHECK_VERSION(3, 0, 0)
+    /* FIXME: apps become invisible on GTK+ 3.*/
     reload_notify_id = menu_cache_add_reload_notify( menu_tree, on_menu_tree_changed, NULL );
+#endif
 
     create_notebook_pages();
 
-- 
2.4.6

